composer/callbacks/eval_output_logging_callback.py

```diff
@@ -0,0 +1,148 @@
+        self.hash.update((str(time.time()) + str(random.randint(0, 1_000_000))).encode('utf-8'))
+        tmp_dir = os.getcwd() + '/' + self.hash.hexdigest()
+
+        if not os.path.exists(tmp_dir):
+            with dist.local_rank_zero_download_and_wait(tmp_dir):
+                if dist.get_local_rank() == 0:
+                    os.mkdir(tmp_dir)
+
+        full_df = pd.DataFrame()
+        file_name = f'eval-outputs-ba{state.timestamp.batch.value}.tsv'
+
+        for benchmark in self.table:
+            cols, rows = self.table[benchmark]
+            rows = [[e.encode('unicode_escape') if isinstance(e, str) else e for e in row] for row in rows]
+            df = pd.DataFrame.from_records(data=rows, columns=cols)
+            df['benchmark'] = benchmark
+            full_df = pd.concat([full_df, df], ignore_index=True)
+
+        with dist.local_rank_zero_download_and_wait(f'{tmp_dir}/{file_name}'):
+            if dist.get_local_rank() == 0:
+                with open(f'{tmp_dir}/{file_name}', 'wb') as f:
+                    full_df.to_csv(f, sep='\t', index=False)
+
+        # copy/upload tmp files
+        _write(destination_path=f'{self.output_directory}/{file_name}', src_file=f'{tmp_dir}/{file_name}')
+        os.remove(f'{tmp_dir}/{file_name}')
+        self.destination_file = f'{self.output_directory}/{file_name}'
+
+        # delete tmp files
+        os.rmdir(tmp_dir)
+
+    def _prep_response_cache(self, state, cache):
+        benchmark = state.dataloader_label
+        for metric in state.eval_metrics[benchmark].values():
+            if hasattr(metric, 'reset_response_cache'):
+                metric.reset_response_cache(cache)
+
+    def eval_start(self, state: State, logger: Logger) -> None:
+        # eval start runs before each benchmark's evaluator (either in training or eval)
+        self._prep_response_cache(state, True)
+
+    def eval_after_all(self, state: State, logger: Logger) -> None:
+        # eval after all runs after all evaluators have completed during eval within training
+        #  (either in training or eval)
+        self._write_tables_to_output_dir(state)
+        self.table = {}
+
+    def eval_standalone_end(self, state: State, logger: Logger) -> None:
+        # eval standalone end runs after all evaluators have completed during a direct call to trainer.eval()
+        self._write_tables_to_output_dir(state)
+        self.table = {}
+
+    def eval_end(self, state: State, logger: Logger) -> None:
+        # eval start runs after each benchmark's evaluator
+        # during each eval, only a single dataloader/benchmark will be active
+        assert state.dataloader is not None
+        assert isinstance(state.dataloader, DataLoader)
+        if hasattr(state.dataloader, 'dataset') and isinstance(state.dataloader.dataset, ICLDatasetTypes):
+            assert isinstance(state.dataloader.dataset, ICLDatasetTypes)
+            if hasattr(state.dataloader.dataset, 'tokenizer'):
```

Comment:
Do datasets ever not have tokenizers?

Reply:
They always do, but this is necessary for pyright to pass